### ISSUE TITLE
Experiment with distributed upload locks

### ISSUE THREAD

**Issue by Acconut** [tus/tusd#917]
Currently we only support memory and file locks, whose reach is limited to a single machine. If you want to properly scale horizontally, a distributed lock is needed, such as Redis, Consul, etcd etc.

@Murderlon proposed the use of https://github.com/matryer/vice, which wraps multiple messaging services into a single interface. Maybe that is helpful. More infos at https://medium.com/@matryer/introducing-vice-go-channels-across-many-machines-bcac1147d7e2.

**Comment by Acconut** [comment]
Thanks for the feedback! RedLock does look interesting, so let me know if you have some more insights.

**Comment by gedw99** [comment]
Also Nats has this ability and it can be embedded.

the lock is just a kv in nats .

kine which is an etcd cluster also uses nats to a hieve global state.

**Comment by Acconut** [comment]
Thanks, NATS does also look interesting on its own.

**Comment by gedw99** [comment]
@Acconut 

https://github.com/maxpert/marmot is a good starter.

It's designed to do data replication between SQLITE databases, but the concepts apply for tracking locks etc. 
It also embedded the NATS Server if you want. 

The other way is that NATS has an Object Store, that can hold the chunks per user upload.
When the upload is done, then you can extract it all out.
https://github.com/a-h/natsjson uses the KV, which is a lot like using the Object Store.

**Comment by Acconut** [comment]
Interesting links, thank you. Let us know if you are interested in trying out NATS for tusd's locks and we can assist you :)

**Comment by gedw99** [comment]
I am interested because I want to use it with Pocketbase, so that uploads can be  resilient

https://github.com/maxpert/marmot/issues/101 Is a brainstorm on doing it with NATS. It already does SQLite replication and file replication is easy to add.

The upload aspect is also handled there but I don’t know if the tusd protocol is conflating things tbh.

**Comment by Acconut** [comment]
> I am interested because I want to use it with Pocketbase, so that uploads can be resilient

Great idea!



> The upload aspect is also handled there but I don’t know if the tusd protocol is conflating things tbh.

Not sure what you mean by this. Do you want to handle file uploads directly via NATS? If so, the tusd might not be necessary for you.

**Comment by gedw99** [comment]
> > I am interested because I want to use it with Pocketbase, so that uploads can be resilient
> 
> Great idea!
> 
> > The upload aspect is also handled there but I don’t know if the tusd protocol is conflating things tbh.
> 
> Not sure what you mean by this. Do you want to handle file uploads directly via NATS? If so, the tusd might not be necessary for you.

Yes  you got it. NATS on the client side will chunk the file ( or any data)  up to the NATS server. If the client goes down , when it comes back up it will ensure the file or data eventually gets to the NATS server. Once it gets to the Server it can then push it to Tusd or anything else.   NATS in this case is acting as a Plane on top of Tusd.  

NATS can currently used with Pocketbase to allow all data transacted to the Pocketbase SQLite DB to be scaled out to all others Pocketbase Sqlite servers.  But its file uploads are scaled out.

**Comment by Acconut** [comment]
Interesting approach, but this issue talks about a different use case for distributed systems. We want to look into how a distributed system can be used to lock upload resources and ensure exclusive access in the presence of multiple requests (see https://github.com/tus/tusd/blob/main/docs/locks.md). What you are proposing is something different AFAIU

**Comment by gedw99** [comment]
thanks for the link @Acconut . I read it pretty quickly and it  makes sense.

"the locks do not extend to every server.". 
You can use NATS to do this so all servers have full knowledge of any LOCK.

See k3 and Kine with uses NATS jet stream to do this also.

NATS will decouple things.

You know the problem domain much better than me however.....

**Comment by cawfeecoder** [comment]
Why wouldn't you do a lease and lock on ETCD?
